###Movie Review Dataset
Well start by loading in the IMDB movie review dataset from keras. This dataset contains 25,000 reviews from IMDB where each one is already preprocessed and has a label as either positive or negative. Each review is encoded by integers that represents how common a word is in the entire dataset. For example, a word encoded by the integer 3 means that it is the 3rd most common word in the dataset.
 



In [1]:
import tensorflow as tf
from tensorflow import keras
from keras.preprocessing import sequence
from keras.datasets import imdb
import os
import numpy as np
VOCAB_SIZE = 88584
max_len=250
batch_size=64

(train_data,train_lables),(test_data,test_lables)=imdb.load_data()
print(tf.__version__)

17473536/17464789 [==============================] - 0s 0us/step
2.8.2


In [2]:
#look at one review
(train_data[1])

[1,
 194,
 1153,
 194,
 8255,
 78,
 228,
 5,
 6,
 1463,
 4369,
 5012,
 134,
 26,
 4,
 715,
 8,
 118,
 1634,
 14,
 394,
 20,
 13,
 119,
 954,
 189,
 102,
 5,
 207,
 110,
 3103,
 21,
 14,
 69,
 188,
 8,
 30,
 23,
 7,
 4,
 249,
 126,
 93,
 4,
 114,
 9,
 2300,
 1523,
 5,
 647,
 4,
 116,
 9,
 35,
 8163,
 4,
 229,
 9,
 340,
 1322,
 4,
 118,
 9,
 4,
 130,
 4901,
 19,
 4,
 1002,
 5,
 89,
 29,
 952,
 46,
 37,
 4,
 455,
 9,
 45,
 43,
 38,
 1543,
 1905,
 398,
 4,
 1649,
 26,
 6853,
 5,
 163,
 11,
 3215,
 10156,
 4,
 1153,
 9,
 194,
 775,
 7,
 8255,
 11596,
 349,
 2637,
 148,
 605,
 15358,
 8003,
 15,
 123,
 125,
 68,
 23141,
 6853,
 15,
 349,
 165,
 4362,
 98,
 5,
 4,
 228,
 9,
 43,
 36893,
 1157,
 15,
 299,
 120,
 5,
 120,
 174,
 11,
 220,
 175,
 136,
 50,
 9,
 4373,
 228,
 8255,
 5,
 25249,
 656,
 245,
 2350,
 5,
 4,
 9837,
 131,
 152,
 491,
 18,
 46151,
 32,
 7464,
 1212,
 14,
 9,
 6,
 371,
 78,
 22,
 625,
 64,
 1382,
 9,
 8,
 168,
 145,
 23,
 4,
 1690,
 15,
 16,
 4,
 1355,
 5,
 28,
 6,
 52,
 

In [3]:
len(train_data)

25000

In [4]:
len(test_data)

25000

In [5]:
len(train_data[1])

189

**More Preprocessing**

If we have a look at some of our loaded in reviews, we'll notice that they are different lengths. This is an issue. We cannot pass different length data into our neural network. Therefore, we must make each review the same length. To do this we will follow the procedure below:

*if the review is greater than 250 words then trim off the extra words

*if the review is less than 250 words add the necessary amount of 0's to make it equal to 250.


In [6]:
train_data= sequence.pad_sequences(train_data,max_len)
test_data=sequence.pad_sequences(test_data,max_len)
len(train_data[1])

250

In [7]:
train_data[1]

array([    0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     1,   194,
        1153,   194,  8255,    78,   228,     5,     6,  1463,  4369,
        5012,   134,    26,     4,   715,     8,   118,  1634,    14,
         394,    20,    13,   119,   954,   189,   102,     5,   207,
         110,  3103,    21,    14,    69,   188,     8,    30,    23,
           7,     4,   249,   126,    93,     4,   114,     9,  2300,
        1523,     5,   647,     4,   116,     9,    35,  8163,     4,
         229,     9,   340,  1322,     4,   118,     9,     4,   130,
        4901,    19,

###Creating the Model
Now it's time to create the model. We'll use a word embedding layer as the first layer in our model and add a LSTM layer afterwards that feeds into a dense node to get our predicted sentiment. 

32 stands for the output dimension of the vectors generated by the embedding layer. We can change this value if we'd like!

In [8]:
model=tf.keras.Sequential([
    tf.keras.layers.Embedding(VOCAB_SIZE,32),
    tf.keras.layers.LSTM(32),
    tf.keras.layers.Dense(1,activation='sigmoid')
])

In [9]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 32)          2834688   
                                                                 
 lstm (LSTM)                 (None, 32)                8320      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 2,843,041
Trainable params: 2,843,041
Non-trainable params: 0
_________________________________________________________________


**TRAINING**

In [10]:
model.compile(loss="binary_crossentropy",optimizer="adam",metrics=['acc'])
history=model.fit(train_data,train_lables,epochs=10,validation_split=0.2)

Epoch 1/10
625/625 [==============================] - 15s 13ms/step - loss: 0.4102 - acc: 0.8123 - val_loss: 0.4048 - val_acc: 0.8198
Epoch 2/10
625/625 [==============================] - 8s 13ms/step - loss: 0.1756 - acc: 0.9383 - val_loss: 0.3052 - val_acc: 0.8802
Epoch 3/10
625/625 [==============================] - 8s 12ms/step - loss: 0.0821 - acc: 0.9738 - val_loss: 0.4365 - val_acc: 0.8720
Epoch 4/10
625/625 [==============================] - 8s 12ms/step - loss: 0.0494 - acc: 0.9848 - val_loss: 0.4891 - val_acc: 0.8712
Epoch 5/10
625/625 [==============================] - 8s 12ms/step - loss: 0.0532 - acc: 0.9833 - val_loss: 0.5170 - val_acc: 0.8636
Epoch 6/10
625/625 [==============================] - 8s 13ms/step - loss: 0.0322 - acc: 0.9904 - val_loss: 0.5283 - val_acc: 0.8298
Epoch 7/10
625/625 [==============================] - 8s 13ms/step - loss: 0.0293 - acc: 0.9912 - val_loss: 0.5193 - val_acc: 0.8676
Epoch 8/10
625/625 [==============================] - 8s 12ms/step -

And we'll evaluate the model on our training data to see how well it performs.

In [11]:
result=model.evaluate(test_data,test_lables)
print(result)

782/782 [==============================] - 5s 6ms/step - loss: 0.7051 - acc: 0.8438
[0.705141007900238, 0.8438000082969666]


In [12]:
word_index=imdb.get_word_index()

def encode(text):
  tokens= keras.preprocessing.text.text_to_word_sequence(text)
  l=[]
  for words in tokens:
    if words in word_index:
      l.append(word_index[words])
    else:
      l.append('0')
  return(sequence.pad_sequences([l],max_len))
text = "that movie was just amazing, so amazing"
encoded = encode(text)
print(encoded)

1654784/1641221 [==============================] - 0s 0us/step
[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0

In [16]:
# now time to make a prediction

def predict(text):
  encoded_text=encode(text)
  pred=np.zeros((1,250))
  pred[0]=encoded_text
  result=model.predict(pred)
  print(result[0])

positive_review = "That movie was! really loved it and would great watch it again because it was amazingly great"
predict(positive_review)

negative_review = "that movie really sucked. I hated it and wouldn't watch it again. Was one of the worst things I've ever watched"
predict(negative_review)

[0.9990275]
[0.5813271]


since we got the movie review prediction for the positive_review as 0.9990275  means nearly 99% rating which says the movie is good to watch...

In the negative review if predicted  the rating as 58% which says that the movie is really not such great to watch